### PR TITLE
Bean Validation

### PR DESCRIPTION
**Pull request by graemerocher**
This PR makes Hibernate Validator an optional dependency and adds the following features:

* Light weight, reflection free validator with minimal startup time impact.
* Compile time validation of the AST and support for using `javax.validation` annotations on other annotation members
* Reactive validation

Documentation will follow in a separate PR once this is merged.

### CODE REVIEW

File: inject-groovy/src/main/groovy/io/micronaut/ast/groovy/InjectTransform.groovy [hunk 1]
```diff
@@ -834,6 +844,9 @@ class InjectTransform implements ASTTransformation, CompilationUnitAware {
                         }
                     } else {
                         Object[] interceptorTypeReferences = methodAnnotationMetadata.getAnnotationNamesByStereotype(Around).toArray()
+                        if (hasConstraints) {
+                            interceptorTypeReferences = ArrayUtils.concat(interceptorTypeReferences, "io.micronaut.validation.Validated")
```
Comment (jameskleeh) [inline inject-groovy/src/main/groovy/io/micronaut/ast/groovy/InjectTransform.groovy]: Why is this necessary given https://github.com/micronaut-projects/micronaut-core/pull/1526/files#diff-08f2cbd92c48cc377b2cdd189badf286R817 ?

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java [hunk 2]
```diff
@@ -2214,6 +2214,18 @@ private void readAllBeanDefinitionClasses() {
         disabled.clear();
 
         initializeEventListeners();
+        if (containsBean(MessageSource.class)) {
```
Comment (jameskleeh) [inline inject/src/main/java/io/micronaut/context/DefaultBeanContext.java]: Is this necessary? Can't you inject `@Secondary` instances into the composite primary bean?

File: validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java [hunk 3]
```diff
@@ -71,26 +94,73 @@ public int getOrder() {
...
+                    return validateReturnValue(executableValidator, context, target, targetMethod, result);
                 } else {
                     throw new ConstraintViolationException(constraintViolations);
                 }
             }
 
         }
     }
+
+    private Object validateReturnValue(@Nonnull ExecutableValidator validator, MethodInvocationContext context, Object target, Method targetMethod, Object result) {
```
Comment (jameskleeh) [inline validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java]: Does this call the method?

File: validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java [hunk 4]
```diff
@@ -46,22 +57,34 @@
...
-                .orElse(null);
+    /**
+     * Creates ValidatingInterceptor from the validatorFactory.
+     *
+     * @param micronautValidator The micronaut validator use if no factory is available
+     * @param validatorFactory   Factory returning initialized {@code Validator} instances
+     */
+    @Inject
+    public ValidatingInterceptor(
+            Validator micronautValidator,
```
Comment (jameskleeh) [inline validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java]: Looks like the validator can be disabled with config. Should this be nullable?

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java (hunk 2 above)
Comment (graemerocher) [reply inject/src/main/java/io/micronaut/context/DefaultBeanContext.java]: Yeah maybe not necessary

File: validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java (hunk 4 above)
Comment (graemerocher) [reply validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java]: Yes it should

File: validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java (hunk 3 above)
Comment (graemerocher) [reply validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java]: Nope
